# A Single Neuron
<https://nnfs.io/bkr/>

In [1]:
inputs =[1, 2, 3]
weights = [0.2, 0.8, -0.5]
bias = 2.0

output = inputs[0] * weights[0] + inputs[1] * weights[1] + inputs[2] * weights[2] + bias
print(output)

2.3


# A Layer of Neurons
<https://nnfs.io/mxo/>
    

In [3]:
inputs = [1, 2, 3, 2.5]

weights1 = [0.2, 0.8, -0.5, 1]
weights2 = [0.5, -0.91, 0.26, -0.5]
weights3 = [-0.26, -0.27, 0.17, 0.87]

bias1 = 2
bias2 = 3
bias3 = 0.5

output = [
    inputs[0] * weights1[0] + inputs[1] * weights1[1] + inputs[2] * weights1[2] + inputs[3] * weights1[3] + bias1,
    inputs[0] * weights2[0] + inputs[1] * weights2[1] + inputs[2] * weights2[2] + inputs[3] * weights2[3] + bias2,
    inputs[0] * weights3[0] + inputs[1] * weights3[1] + inputs[2] * weights3[2] + inputs[3] * weights3[3] + bias3
]
print(output)

[4.8, 1.21, 2.385]


In [4]:
weights = [
    weights1,
    weights2,
    weights3
]

biases = [bias1, bias2, bias3]

# Output of current layer
layer_outputs = []
# For each neuron
for neuron_weights, neuron_bias in zip(weights, biases):
    # Zeroed output of given neuron
    neuron_output = 0
    # For each input and weight to neuron
    for n_input, weight in zip(inputs, neuron_weights):
        # Multiply this input by associated weight
        # and add to the neuron's output variable
        neuron_output += n_input * weight
    # Add bias
    neuron_output += neuron_bias
    # Put neuron's result to the layer's output list
    layer_outputs.append(neuron_output)

print(layer_outputs)

[4.8, 1.21, 2.385]


# Tensors、Arrays、Vectors
> A tensor object is an object that can be represented as an array

What this means is, as programmers, we can (and will) treat tensors as arrays in the context of
deep learning, and that’s really all the thought we have to put into it. Are all tensors ​just​ arrays?
No, but they are represented as arrays in our code, so, to us, they’re only arrays, and this is why
there’s so much argument and confusion

Now, what is an array? In this book, we define an array as an ordered homologous container for
numbers, and mostly use this term when working with the NumPy package since that’s what the
main data structure is called within it. A linear array, also called a 1-dimensional array, is the
simplest example of an array, and in plain Python, this would be a list. Arrays can also consist
of multi-dimensional data, and one of the best-known examples is what we call a matrix in
mathematics, which we’ll represent as a 2-dimensional array. Each element of the array can be
accessed using a tuple of indices as a key, which means that we can retrieve any array element.

https://nnfs.io/jps/

In [9]:
# Everything shown so far could also be an array or an array representation of a tensor. 
l = [1,5,6,2]
lol = [[1,5,6,2],
      [3,2,1,3]]

# 3-dimensional array
lolol = [[[1,5,6,2],
          [3,2,1,3]],
          [[5,2,1,2],
           [6,4,8,4]],
         [[2,8,5,3],
         [1,1,9,4]]]
print(l)
print(lol)
print(lolol)
print(lolol[0])

[1, 5, 6, 2]
[[1, 5, 6, 2], [3, 2, 1, 3]]
[[[1, 5, 6, 2], [3, 2, 1, 3]], [[5, 2, 1, 2], [6, 4, 8, 4]], [[2, 8, 5, 3], [1, 1, 9, 4]]]
[[1, 5, 6, 2], [3, 2, 1, 3]]


## 不是Array

In [ ]:
another_list_of_lists = [[4,2,4],
                        [5,1]]

The above list of lists cannot be an array because it is not **homologous**.

A list of lists is homologous if each list along a dimension is identically long, and this must be true for each dimension.

## Matrix

A matrix is pretty simple. It’s a rectangular array. It has columns and rows. It is two dimensional.
So a matrix can be an array (a 2D array). Can all arrays be matrices? No. An array can be far
more than just columns and rows, as it could have four dimensions, twenty dimensions, and so on.

The first dimension is what’s inside the most outer brackets, 

The next dimension’s size is the number of elements inside this more inner pair of brackets

## Shape

In [11]:
# 3x2 or (3,2)
list_matrics_array = [[4,2],
               [5,1],
               [8,2]]

# Vector
a vector in math is what we call a list in Python or a 1-dimensional array in NumPy

在计算机科学和数学领域，Array（数组）、Matrix（矩阵）、Vector（向量）和Tensor（张量）是不同的数据结构或数学概念，它们在不同的领域和上下文中有着不同的意义和用途。

**数组（Array）：**
+ 数组是一种数据结构，由一组相同类型的元素组成，这些元素按照顺序排列并在内存中连续存储。
+ 数组的长度是固定的，一旦创建后大小通常不能改变。
+ 在计算机编程中，数组可以是一维的、多维的或者动态扩展的数据结构。

**矩阵（Matrix）：**
+ 矩阵是数学概念，是一个二维的、由行和列组成的数据结构。它是线性代数和数学运算中的基本概念之一。
+ 矩阵在数学、物理学、工程学等领域有广泛的应用，特别是在涉及到线性变换、解方程组、图像处理等方面。

**向量（Vector）：**
+ 向量是矩阵的一个特殊情况，通常是一个单列的矩阵或一个一维的数组。
+ 在数学上，向量表示为有序排列的一系列数值。在几何学中，它通常代表空间中的方向和大小。
+ 向量在物理学、工程学、计算机图形学等领域有广泛的应用，用于描述和处理空间中的方向、速度、力等概念。

**张量（Tensor）：**
+ 张量是一个广义的概念，它可以表示多维数组，是数组和矩阵的扩展。张量可以是零维的（标量）、一维的（向量）、二维的（矩阵），以及更高维度的数组。
+ 在数学、物理学、机器学习和深度学习等领域中，张量是非常重要的概念。在机器学习和深度学习中，神经网络处理的数据通常以张量的形式表示。

在机器学习和深度学习领域，特别是使用深度神经网络时，张量成为了一个核心概念，因为神经网络处理的输入、输出和中间层数据都以张量的形式进行表示和处理。 TensorFlow 和 PyTorch 等深度学习框架提供了针对张量操作的丰富功能，使得对张量进行高效操作和运算变得非常便捷。







# dot product
$\vec{a}\cdot\vec{b}$  = $\displaystyle \sum_{i = 1}^{n}{a_{i}b_{i}}$ = $a_1b_1 + a_2b_2 + ... + a_nb_n$

A dot product of two vectors is a sum of products of consecutive vector elements

https://nnfs.io/xpo/

另：
$\vec{a}+ \vec{b}$ = [$a_1+b_1 , a_2+b_2 , ... , a_n+b_n$]

In [13]:
# a as inputs , b as weights
a = [1, 2, 3]
b = [2, 3, 4]

dot_product = a[0] * b[0] + a[1] * b[1] + a[2] * b[2]
print(dot_product)

20


# A Single Neuron with Numpy

<https://nnfs.io/blq/>

In [14]:
import numpy as np

inputs = [1.0, 2.0, 3.0, 2.5]
weights = [0.2, 0.8, -0.5, 1.0]
bias = 2.0

outputs = np.dot(weights, inputs) + bias
print(outputs)

4.8


# A Layer of Neurons with Numpy

<https://nnfs.io/cyx/>

In [15]:
import numpy as np

inputs = [1.0, 2.0, 3.0, 2.5]
weights = [[0.2, 0.8, -0.5, 1],
           [0.5, -0.91, 0.26, -0.5],
           [-0.26, -0.27, 0.17, 0.87]]
biases = [2.0, 3.0, 0.5]

layer_outputs = np.dot(weights, inputs) + biases
print(layer_outputs)

[4.8   1.21  2.385]


# A Batch of Data

inputs被称作

**feature set instance**

**observation**

**sample**

<https://nnfs.io/lqw/>

One reason is that it’s faster to train in batches in parallel processing, and the other reason is that batche help with generalization during training. 

https://nnfs.io/vyu/

In [16]:
inputs = [[1, 2, 3, 2.5], [2, 5, -1, 2], [-1.5, 2.7, 3.3, -0.8]]

# Matrix Product

https://nnfs.io/jei/

To perform a matrix product, the size of the second dimension of the left matrix must match the
size of the first dimension of the right matrix. For example, if the left matrix has a shape of ​(5, 4)
then the right matrix must match this 4 within the first shape value ​(4, 7)​. The shape of the
resulting array is always the first dimension of the left array and the second dimension of the right
array, ​(5, 7)​. 

[row and column vector]<https://nnfs.io/bkw/>

# Transposition for the Matrix Product

$\vec{a}\cdot\vec{b}$ =  $ab^\mathrm T$ = $\left[ \begin{matrix} 1 &2 & 3\end{matrix} \right]$ $\left[ \begin{matrix} 2\\3\\4\end{matrix} \right]$ = [20]

https://nnfs.io/qut/

A row vector is a matrix whose first dimension’s size (the number of rows) equals 1 and the
second dimension’s size (the number of columns) equals ​n​ — the vector size. In other words, it’s
a 1×n array or array of shape (1, n)

a = $\left[ \begin{matrix} a_1 &a_2 &a_3 &\cdots &a_n\end{matrix} \right]$

In [19]:
np.array([[1,2,3]])

array([[1, 2, 3]])

In [20]:
a = [1, 2, 3]
np.array([a])

array([[1, 2, 3]])

In [21]:
np.expand_dims(np.array(a), axis=0)

array([[1, 2, 3]])

A column vector is a matrix where the second dimension’s size equals 1, in other words, it’s an
array of shape (n, 1):

b = $\left[ \begin{matrix} b_1 \\b_2 \\b_3\\ \vdots \\b_n\end{matrix} \right]$

$\left[ \begin{matrix} b_1 &b_2 &b_3 &\cdots &b_n\end{matrix} \right]^\mathrm T$ =  $\left[ \begin{matrix} b_1 \\b_2 \\b_3\\ \vdots \\b_n\end{matrix} \right]$

$\left[ \begin{matrix} b_1 \\b_2 \\b_3\\ \vdots \\b_n\end{matrix} \right]^\mathrm T$ = $\left[ \begin{matrix} b_1 &b_2 &b_3 &\cdots &b_n\end{matrix} \right]$

In [2]:
import numpy as np
a = [1,2,3]
b = [2,3,4]

a = np.array([a])
b = np.array([b]).T

np.dot(a, b)

array([[20]])

# A Layer of Neurons & Batch of Data w/Numpy

https://nnfs.io/crq/

In [4]:
import numpy as np
inputs = [[1.0, 2.0, 3.0, 2.5],
          [2.0, 5.0, -1.0, 2.0],
          [-1.5, 2.7, 3.3, -0.8]]

weights = [[0.2, 0.8, -0.5, 1.0],
           [0.5, -0.91, 0.26, -0.5],
           [-0.26, -0.27, 0.17, 0.87]]

print(np.dot(inputs, weights))

ValueError: shapes (3,4) and (3,4) not aligned: 4 (dim 1) != 3 (dim 0)

In [16]:
print(np.dot(inputs, np.array(weights).T))
np.array(weights).T

[[ 2.8   -1.79   1.885]
 [ 6.9   -4.81  -0.3  ]
 [-0.59  -1.949 -0.474]]


array([[ 0.2 ,  0.5 , -0.26],
       [ 0.8 , -0.91, -0.27],
       [-0.5 ,  0.26,  0.17],
       [ 1.  , -0.5 ,  0.87]])

$\left[ \begin{matrix}0.2 &0.8 & -0.5 & 1.0 \\
               0.5 &-0.91 & 0.26 & -0.5 \\
               -0.26 & -0.27 & 0.17 & 0.87\end{matrix} \right]^\mathrm T= 
\left[\begin{matrix} 0.2 & 0.5 & -0.26\\
               0.8 & -0.91 & -0.27\\
               -0.5 & 0.26 & 0.17 \\
               1.0 & -0.5 & 0.87 \end{matrix}\right]$              
               
               
 $inputs \cdot weights^\mathrm T$  $=$      
 $\left[ \begin{matrix}1.0 &2.0 & 3.0 & 2.5 \\
                2.0 &5.0 & -1.0& 2.0 \\
                -1.5 & 2.7 & 3.3 & -0.8\end{matrix} \right]$ 
 $\left[\begin{matrix} 0.2 & 0.5 & -0.26\\
               0.8 & -0.91 & -0.27\\
               -0.5 & 0.26 & 0.17 \\
               1.0 & -0.5 & 0.87 \end{matrix}\right]$ $=$
 $\left[\begin{matrix}
1.0*0.2 + 2.0*0.8 + 3.0*-0.5 + 2.5*1.0 & 1.0*0.5 + 2.0*-0.91 + 3.0*0.26 + 2.5*-0.5 & 1.0*-0.26 + 2.0*-0.27 + 3.0*0.17 + 2.5*0.87\\
2.0*0.2 + 5.0*0.8 + -1.0*-0.5 + 2.0*1.0 & 2.0*0.5 + 5.0*-0.91 + -1.0*0.26 + 2.0*-0.5 & 2.0*-0.26 + 5.0*-0.27 + -1.0 * 0.17 + 2.0*0.87\\
-1.5*0.2 + 2.7*0.8 + 3.3*-0.5 + -0.8*1.0 & -1.5*0.5 + 2.7*-0.91 + 3.3*0.26 + -0.8*-0.5 & -1.5*-0.26 + 2.7*-0.27 + 3.3*0.17 + -0.8*0.87
 \end{matrix}\right]$ $=$

$\left[\begin{matrix}2.8&-1.79&1.885\\
              6.9&-4.81&-0.3\\
              -0.59&-1.949&-0.474\end{matrix}\right]$
              
https://nnfs.io/gjw/

In [ ]:
# 1.0*0.2 + 2.0*0.8 + 3.0*-0.5 + 2.5*1.0
# 1.0*0.5 + 2.0*-0.91 + 3.0*0.26 + 2.5*-0.5
# 1.0*-0.26 + 2.0*-0.27 + 3.0*0.17 + 2.5*0.87
# 2.0*0.2 + 5.0*0.8 + -1.0*-0.5 + 2.0*1.0
# 2.0*0.5 + 5.0*-0.91 + -1.0*0.26 + 2.0*-0.5
# 2.0*-0.26 + 5.0*-0.27 + -1.0 * 0.17 + 2.0*0.87
# -1.5*0.2 + 2.7*0.8 + 3.3*-0.5 + -0.8*1.0
# -1.5*0.5 + 2.7*-0.91 + 3.3*0.26 + -0.8*-0.5
# -1.5*-0.26 + 2.7*-0.27 + 3.3*0.17 + -0.8*0.87

In [52]:
biased = [2.0, 3.0, 0.5]
outputs = np.dot(inputs, np.array(weights).T) + biased
print(outputs)
# https://nnfs.io/qty/

[[ 4.8    1.21   2.385]
 [ 8.9   -1.81   0.2  ]
 [ 1.41   1.051  0.026]]


In [53]:
import numpy as np
inputs = [[1.0, 2.0, 3.0, 2.5],
          [2.0, 5.0, -1.0, 2.0],
          [-1.5, 2.7, 3.3, -0.8]]

weights = [[0.2, 0.8, -0.5, 1.0],
           [0.5, -0.91, 0.26, -0.5],
           [-0.26, -0.27, 0.17, 0.87]]

biased = [2.0, 3.0, 0.5]
layer_outputs = np.dot(inputs, np.array(weights).T) + biased
print(layer_outputs)

[[ 4.8    1.21   2.385]
 [ 8.9   -1.81   0.2  ]
 [ 1.41   1.051  0.026]]
